src/charm.py

```diff
@@ -35,83 +28,83 @@ def __init__(self, *args: typing.Any):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
-        """Define and start a workload using the Pebble API.
+        self._cloudflare_route = CloudflaredRouteProvider(charm=self)
+        self._ingress = IngressPerAppProvider(charm=self)
+        self.framework.observe(self.on.config_changed, self._reconcile)
+        self.framework.observe(self.on.secret_changed, self._reconcile)
+        self.framework.observe(self._ingress.on.data_provided, self._reconcile)
+        self.framework.observe(self.on["cloudflared-route"].relation_changed, self._reconcile)
+        self.framework.observe(self.on.get_ingress_data_action, self._on_get_ingress_data_action)
+
+    def _reconcile(self, _: ops.EventBase) -> None:
+        """Handle changed configuration."""
+        if not self.unit.is_leader():
+            self.unit.status = ops.BlockedStatus(
+                "this charm only supports a single unit, please remove the additional units "
+                f"using `juju scale-application {self.app.name} 1`"
+            )
+            return
+        self.unit.status = ops.ActiveStatus()
+        domain = self.config.get("domain")
+        tunnel_token = self._get_tunnel_tokens()
+        if not (domain and tunnel_token):
+            missing = []
+            if not domain:
+                missing.append("domain")
+            if not tunnel_token:
+                missing.append("tunnel-token")
+            self.unit.status = ops.BlockedStatus(f"waiting for {', '.join(missing)} configuration")
+            self._unpublish_ingress_url()
+            return
+        if relation := self.model.get_relation("cloudflared-route"):
+            self._cloudflare_route.set_tunnel_token(tunnel_token, relation=relation)
+            for ingress_relation in self._ingress.relations:
```

Comment:
There can be only one, do we really need to iterate ?

Reply:
I guess we should block if there are multiple as well

Reply:
There will only be at most one relation, I changed it to the equivalent form:
```
if self._ingress.relations:
    self._ingress.publish_url(self._ingress.relations[0], f"https://{domain}")
```